atom centered basis set and related metadata definitions

EBB2675 opened the pull request:
Here is the support for `AtomCenteredFunction`, `AtomCenteredBasisSet` and co.

As you know, i follow the TREXIO format:
https://trex-coe.github.io/trexio/trex.html

Meanwhile, Denis will also take a look at it from a scientific perspective.

EBB2675 committed:
migrate atom centered basis set to this branch

EBB2675 committed:
add tests, finalize AtomCenteredFunction and basis set

EBB2675 committed:
remove padding and truncation logic

JFRudzinski commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+
+    # helper (internal)
+    n_components = Quantity(
+        type=np.int32, description='Number of components implied by function_type.'
```
If this is only a helper then should it simply be a private class variable and not a proper quantity, since it seems that after normalization it will simply be = 1?

JFRudzinski commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+
+        # expand combined shells and exit
+        if self._is_combined():
+            self._expand_combined_into_parent()
```
I don't fully understand what is going on here, but wondering if you need to guard against duplicate expansions in case of future refactors, e.g., is this covered in the tests?

EBB2675 replied:
i have removed expansion entirely

JFRudzinski commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        """,
+    )
+
+    ao_custom_order = Quantity(
```
have you considered `type=JSON`?

EBB2675 replied:
done

JFRudzinski commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
     """
 
-    pass
+    harmonic_type = Quantity(
```
is "cartesian harmonic" correct terminology? Maybe rename "angular_type" or "angular_representation". Enum types could be expanded to `spherical_harmonics` / `cartesian_polynomials` (or whatever is accurate terminology) if necessary to be clear

JFRudzinski reviewed (approved):
:1st_place_medal:

ladinesa commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        instance in place to the first component and appending the rest to the
+        parent's functional_compositions. Quiet no-op if parent/data missing.
+        """
+        parent = getattr(self, 'm_parent', None)
```
do not use getattr parent = self.m_parent

ladinesa commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        parent's functional_compositions. Quiet no-op if parent/data missing.
+        """
+        parent = getattr(self, 'm_parent', None)
+        if parent is None or not hasattr(parent, 'functional_compositions'):
```
```
if not isinstance(parent, AtomCenteredBasisFunctions):
    return
```

ladinesa commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+
+        # check expected length consistency, bail out if mismatch
+        if flat.size != expected:
+            return
```
log error.

ladinesa commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,521 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        if flat.size != expected:
+            return
+
+        # ---- rewrite THIS instance to the first component ----
```
maybe try to put this in the loop then say shell = AtomCenteredfunction() if n else self

EBB2675 committed:
fix according to reviews

EBB2675 committed:
fix AttributeError in AtomCenteredFunction

EBB2675 committed:
simplify handling combined shells

EBB2675 committed:
remove getattr and hasattr

EBB2675 committed:
remove excessively detailed primitive_normalization quantity

EBB2675 committed:
avoid second-guessing ECP family name

EBB2675 commented:
Hi again @ladinesa @JFRudzinski 

I've pushed the final commits for the atom-centered basis. The PR is ready for final quick review.

- The main changes resolve the complexity around combined shells:  I removed the ambiguous combined shell types ('sp', 'spd', etc.) and all associated expansion/deletion logic. The schema now only accepts single-ℓ shells, shifting the expansion task to the parser (which is cleaner).

- I also eliminated all remaining unnecessary `getattr` calls.

- I anticipate applying more granular consistency checks and physical constraints in the future via normalization as I implement more parsers. But this shouldn’t stop us from merging this I think.

- Of course, I added a lot of tests, e.g. a mixed-atom centered setup where different basis sets were assigned to different atomic species and so on.

JFRudzinski reviewed (approved):
Looks good to me :+1:

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+
+    In many quantum-chemistry codes, an atom-centered basis set is composed of
+    several "shells," each shell containing one or more basis functions of a certain
+    angular momentum. For instance, a shell of p-type orbitals (L=1) typically
```
...(ℓ=1)...

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    A single "atom-centered function" can be a linear combination of multiple
+    primitive Gaussians (or Slater-type orbitals, STOs).
+    In practice, these contract together to form the final basis function used by
+    the SCF or post-SCF method. Often, each contraction is labeled by its
```
Mm I am a bit confused about what the angular momentum ℓ then represents. I guess "contraction" means here "degeneracy", i.e., that for same values of ℓ there are multiple mℓ?

Just to clarify for me.

EBB2675 replied:
danke sehr, I ll update the docs to make it clear:

ℓ is the shell angular momentum as usual. A contracted AO is a linear combination of primitives with the same center & ℓ; the degeneracy comes from the 2ℓ+1 components

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    )
+
+    # explicit single-ℓ metadata
+    angular_momentum = Quantity(
```
`function_type` and `angular_momentum` are redundant (right?). If so, can you mention this in the description?

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    )
+
+    # per-shell normalization (contracted)
+    shell_normalization = Quantity(
```
is this without units? not sure from the description what is the normalization here (does it mean degeneracy-related normalization?)

EBB2675 replied:
yes, this quantity is unitless.
It represents the normalization factor applied to each contracted AO (shell) so that the function satisfies the chosen normalization convention (e.g., each AO integrates to 1 over all space).

It’s not related to degeneracy, but to how the primitive Gaussians are combined and scaled. For example, if primitives are individually normalized and then combined, a further contraction normalization may be required to ensure the final AO remains normalized.

I’ve clarified the description to reflect this :+1:

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+
+    # primitives
+    n_primitive = Quantity(
+        type=np.int32, description='Number of primitives in this shell.'
```
maybe mention again briefly what a primitive is. Either here or in `exponents`

EBB2675 replied:
done

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        type=np.int32, description='Number of primitives in this shell.'
+    )
+
+    exponents = Quantity(
```
I would keep naming here with `_primitive` to make clear this is related with the expansion in gaussians/whatever

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    )
 
-    # TODO: design system for writing basis functions like gaussian or slater orbitals
+    primitive_factor = Quantity(
```
you can keep consistent naming with `..._primitive` (or `primitive_...`

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        type=np.float32, description='Optional embedded point charge.'
+    )
+
+    _L_MAP = {
```
I don't know how many times we wrote this mapping... 😄 I had: https://github.com/FAIRmat-NFDI/nomad-parser-wannier90/blob/develop/src/nomad_parser_wannier90/parsers/win_parser.py#L60

@ladinesa @ndaelman-hu would love if we could formalize this somewhere in the package, separated from here.

EBB2675 replied:
im adding an issue about this 👍

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    """
+
+    num = Quantity(type=np.int32, description='Total number of AOs (contracted).')
+    cartesian = Quantity(type=bool, description='True=Cartesian, False=Spherical.')
```
a bit inconsistent with the MEnum above in AtomCenteredFunction

I would keep the type as a MEnum

EBB2675 replied:
done 👍

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        population analysis, MO coefficients, or density matrices.
+    """
+
+    num = Quantity(type=np.int32, description='Total number of AOs (contracted).')
```
`n_atomic_orbitals`

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        shape=['num'],
+        description="For each AO: index of AtomCenteredFunction (the 'shell') it belongs to.",
+    )
+    normalization = Quantity(
```
yeah, still confused also when thinking about `shell_normalization` from above.

EBB2675 replied:
elaborated in the docstring

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        """,
+    )
+
+    ao_ordering_convention = Quantity(
```
Mmm not 100% happy with this, as the order should be decided by us (and parsing takes care of putting it in the correct format)

EBB2675 replied:
even if parsing takes care of the canonicalization, it is common practice to keep track of the custom order if given

JosePizarro3 commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        """,
+    )
+
+    total_number_of_basis_functions = Quantity(
```
`n_total_basis_functions`

JosePizarro3 reviewed (approved):
Sorry for my late review, the notification ended up deep in my inbox

I like the implementation. Also reminded me that eventually I need to go through this for Wannier and Slater-Koster models.

ndaelman-hu commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    the SCF or post-SCF method. Often, each contraction is labeled by its
+    angular momentum (e.g., s, p, d, f) and a set of exponents and coefficients.
+
+    This class also accepts *combined* shells ('sp', 'spd', 'spdf') as input.
+    On normalize(), it **expands** them into multiple single-ℓ shells and
+    replaces itself with those shells in the parent's `functional_compositions`.
```
```suggestion
```

> The main changes resolve the complexity around combined shells: I removed the ambiguous combined shell types ('sp', 'spd', etc.) and all associated expansion/deletion logic.

I guess this part of the docs should also be removed?

EBB2675 replied:
ah yes, thank you!

ndaelman-hu commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
     """
-    Defines an atom-centered basis set.
+    Defines an **atom-centered basis set** for quantum chemistry calculations.
+    Unlike plane-wave methods, these expansions are typically built around each atom's
```
```suggestion
    These expansions are typically built around each atom's
```

I wouldn't try to draw comparisons or distinctions between different techniques for mostly different communities. Each can stand on its own.
Btw, in practice PW methods have a dedicated approach for describing the core electrons, be it pseudopotentials or LAPW. PW is just used for describing the valence electrons at a system-level.

ndaelman-hu commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    basis set name (e.g., "cc-pVTZ", "def2-SVP", "6-31G**") and specify the high-level
+    role of the basis set in the calculation.
+
+    **Common examples**:
+      - **Pople basis** (3-21G, 6-31G(d), 6-311++G(2df,2pd), etc.)
+      - **Dunning correlation-consistent** (cc-pVDZ, cc-pVTZ, aug-cc-pVTZ, etc.)
+      - **Slater basis** used in ADF, for instance
+      - **ECP** (Effective Core Potential) expansions like LANL2DZ or SDD for transition metals
+
+    **References**:
+      - F. Jensen, *Introduction to Computational Chemistry*, 2nd ed., Wiley (2007).
+      - A. Szabo, N. S. Ostlund, *Modern Quantum Chemistry*, McGraw-Hill (1989).
+      - T. H. Dunning Jr., J. Chem. Phys. 90, 1007 (1989) for correlation-consistent basis sets.
     """
 
-    functional_composition = SubSection(
+    basis_set = Quantity(
+        type=str,
+        description="""
+        **Name** or label of the basis set as recognized by the code or standard 
+        library. Examples: "6-31G*", "cc-pVTZ", "def2-SVP", "STO-3G", "LANL2DZ" (ECP).
+        """,
+    )
```
I like the flexible approach of both using a standardized name as well as its building blocks.
**Please elaborate on whether the former suffices, or if the latter should always be present.**

EBB2675 replied:
thanks alot, very important point.

**Policy**

- **`basis_set` name alone is sufficient** when the parser can resolve that name to a **unique, versioned definition for all species (including any ECPs)** under the emitting code’s version (e.g., `def2-TZVP`, `cc-pVTZ`, `STO-3G`). In that case, the name is authoritative and the explicit shells are optional.

- **Explicit shell/ECP blocks are required** if any of the following holds:
  1. per-species **mixtures** (e.g., `def2-TZVP` on C/H, `def2-SVP` on N),
  2. **augmented, truncated, or edited** sets (added diffuse, “no-f”, re-contracted, custom exponents/coefficients),
  3. **ambiguous aliases** or code-specific variants that can’t be uniquely mapped,
  4. **partial coverage** (some elements rely on a fallback or custom block),
  5. the code emits AOs that **don’t match** the canonicalized order unless we materialize the explicit shells here for correctness/provenance.

- **Parsers** are responsible for the mapping **code order → NOMAD canonical order** and for deciding whether the name is uniquely resolvable. When in doubt, they should include the explicit blocks **in addition to** the name.

I am adding this policy to the quantity docstring 👍

ndaelman-hu commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+    basis set name (e.g., "cc-pVTZ", "def2-SVP", "6-31G**") and specify the high-level
+    role of the basis set in the calculation.
+
+    **Common examples**:
+      - **Pople basis** (3-21G, 6-31G(d), 6-311++G(2df,2pd), etc.)
+      - **Dunning correlation-consistent** (cc-pVDZ, cc-pVTZ, aug-cc-pVTZ, etc.)
+      - **Slater basis** used in ADF, for instance
+      - **ECP** (Effective Core Potential) expansions like LANL2DZ or SDD for transition metals
+
+    **References**:
+      - F. Jensen, *Introduction to Computational Chemistry*, 2nd ed., Wiley (2007).
+      - A. Szabo, N. S. Ostlund, *Modern Quantum Chemistry*, McGraw-Hill (1989).
+      - T. H. Dunning Jr., J. Chem. Phys. 90, 1007 (1989) for correlation-consistent basis sets.
     """
 
-    functional_composition = SubSection(
+    basis_set = Quantity(
+        type=str,
+        description="""
+        **Name** or label of the basis set as recognized by the code or standard 
+        library. Examples: "6-31G*", "cc-pVTZ", "def2-SVP", "STO-3G", "LANL2DZ" (ECP).
+        """,
+    )
```
Suggestion for future: I'm sure you're already aware of [BSE](https://www.basissetexchange.org/).
This can help bridge the gap between both levels of description.

EBB2675 replied:
yes, just like the Libxc canonicalization in DFT, I wish to do it with BSE for basis sets :+1:

ndaelman-hu commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+                        logger.warning(
+                            'shell_index contains index values outside the bounds of functional_compositions.'
+                        )
+                except Exception:
+                    # Log an error for unexpected exception during max_idx check
+                    logger.error(
+                        'Failed to perform max_idx sanity check on shell_index.'
+                    )
```
Consider Alvin's `log` decorator. You might have to move this part to its own method if you just want to target that.

ndaelman-hu commented on tests/test_basis_set.py:
```diff
@@ -419,3 +422,442 @@ def test_quick_step() -> None:
+# -------------------------------
+
+
+def test_acf_infers_n_primitive_and_ell() -> None:
```
`pytest.mark.parametrize` for testing various edge cases, e.g. `n_primitive == 0` (when the parser doesn't populate the section), etc.

ndaelman-hu replied:
The same comment may apply to some of the other tests below. I leave it up to your discretion which you extend.

ndaelman-hu commented on tests/test_basis_set.py:
```diff
@@ -419,3 +422,442 @@ def test_quick_step() -> None:
+    assert acf.angular_momentum == 2  # d → ℓ=2
+
+
+def test_acf_invalid_length_reset() -> None:
+    """
+    Tests that if n_primitive is set, but coefficient/exponent arrays mismatch length,
+    they are correctly reset to None (due to validation failure).
+    """
```
If this generates any logs, you can check them too. Just provide a test logger object.

ndaelman-hu commented on tests/test_basis_set.py:
```diff
@@ -419,3 +422,442 @@ def test_quick_step() -> None:
+      - ECP only for O
+    Verifies component wiring, roles, species_scope mapping (via m_to_dict), and ECP consistency.
+    """
+    entry = EntryArchive(
+        data=Simulation(
+            model_system=[
+                ModelSystem(
+                    cell=[AtomicCell()],
+                    particle_states=[
+                        AtomsState(
+                            chemical_symbol='H'
+                        ),  # /data/model_system/0/particle_states/0
+                        AtomsState(
+                            chemical_symbol='O'
+                        ),  # /data/model_system/0/particle_states/1
+                    ],
+                )
+            ],
+            model_method=[
+                ModelMethod(
+                    contributions=[
+                        BaseModelMethod(name='kinetic'),
+                        BaseModelMethod(name='electron-ion'),
+                        BaseModelMethod(name='hartree'),
+                    ],
+                    numerical_settings=[],
+                )
+            ],
+        )
+    )
```
This looks like a re-usable setup. I would produce it outside of the method, so here you focus in testing.
If you use a similar template often, consider moving it to the pytest `config`.

ndaelman-hu reviewed (approved):
Good work! Very minor feedback.

EBB2675 committed:
docstring modifications according to reviews

EBB2675 committed:
rename quantities

EBB2675 committed:
add basis set naming convention policy to docstring

EBB2675 committed:
rename quantity

EBB2675 committed:
elaborate primitives

EBB2675 committed:
elaborate AO normalization

EBB2675 committed:
fix tests
